python/cudf/cudf/core/column/column.py

```diff
@@ -489,80 +491,130 @@ def __setitem__(self, key: Any, value: Any):
+                )
+
+        # stride != 1, create a scatter map with arange
+        scatter_map: ColumnBase = arange(
+            start=key_start,
+            stop=key_stop,
+            step=key_stride,
+            dtype=cudf.dtype(np.int32),
+        )
+
+        return self._scatter_by_column(
+            cast("cudf.core.column.NumericalColumn", scatter_map), value
+        )
+
+    def _scatter_by_column(
+        self,
+        key: "cudf.core.column.NumericalColumn",
+        value: Union[cudf.core.scalar.Scalar, ColumnBase],
+    ) -> ColumnBase:
+        nelem: int  # the number of elements to scatter
+
+        if is_bool_dtype(key.dtype):
+            if not len(key) == len(self):
+                raise ValueError(
+                    "Boolean mask must be of same length as column"
+                )
+            nelem = -1
+            if isinstance(value, ColumnBase):
+                if len(self) == len(value):
+                    # Both value and key is aligned to self. Thus, the values
+                    # corresponding to the `F` masks in key should be ignored.
+                    value = value.apply_boolean_mask(key)
+                    nelem = len(value)
+            # For boolean masks, the number of element to scatter is the number
+            # of `true` values in the mask.
+            nelem = key.sum() if nelem == -1 else nelem
         else:
-            key = as_column(key)
-            if is_bool_dtype(key.dtype):
-                if not len(key) == len(self):
-                    raise ValueError(
-                        "Boolean mask must be of same length as column"
-                    )
-                key = arange(len(self))[key]
-                if hasattr(value, "__len__") and len(value) == len(self):
-                    value = as_column(value)[key]
             nelem = len(key)
 
-        if is_scalar(value):
-            value = cudf.Scalar(value, dtype=self.dtype)
-        else:
+        self._check_scatter_key_length(nelem, value)
+
+        try:
+            if is_bool_dtype(key.dtype):
+                # boolean_mask_scatter assigns the ith row of value to the row
+                # where the ith `true` resides in mask.
+                return libcudf.copying.boolean_mask_scatter(
+                    [value], [self], key
+                )[0]._with_type_metadata(self.dtype)
+            else:
+                return libcudf.copying.scatter(
+                    value, key, self
+                )._with_type_metadata(self.dtype)
+        except RuntimeError as e:
+            if "out of bounds" in str(e):
+                raise IndexError(
+                    f"index out of bounds for column of size {len(self)}"
+                ) from e
+            raise
+
+    def _check_scatter_key_length(
+        self, nelem: int, value: Union[cudf.core.scalar.Scalar, ColumnBase]
+    ):
+        """:param nelem: the number of keys to scatter. Should equal to the
+        number of value.
```

Comment:
Wrong style of docstring sneaking in here :) I'd be fine avoiding annotations for such a deeply internal function as long as the description is informative.